### PR TITLE
Fix couple of test bugs, add tests for vis:open() and vis:command_register()

### PR DESCRIPTION
**Pull request by joshaw**
The tests for `vis:open()` are failing because both `vis:open(filename)` and
`vis:command('e filename')` open the file in a split window and do not change
focus to that window. Is this expected? I was expecting the new file to replace
the existing window.

Some of the `command_register` tests are failing, all due to cursor.pos and
range. I know these are hot off the press, but the capability looked
interesting, so thought I'd add some tests. Not sure what the correct values
are supposed to be. Are these byte based?

### CODE REVIEW

**Comment by martanne** [comment]
I will cherry pick the fixes, thanks! As for the `command_register` tests it might be a bit too early for these, as the behavior/API will probably change. I will add some inline comments though.

In general it seems like a good idea to create a "meta" issue in the vis repository named something like "Lua plugin API" where possible API improvements can be discussed etc. Otherwise stuff will be forgotten.

Things that come to mind right now are:
- inconsistency in cursor positioning: `cursor:to(line,col)` should imply `cursor.line == line and cursor.col == col`
- handling of invalid positions: negative or beyond EOF

As for the testing infrastructure, why do we need two different kind of tests? Why can't we just commit the expected `*.status` file as a regular `*.ref` file. The actual content should not matter, i.e. whether we compare a text or a list of booleans makes no difference.

Also it would be nice if the `test.sh` script would avoid bashism. I haven't actually looked too closely at the code, I just noticed that the sheband line is `#!/bin/bash`

File: lua/basic_command_register.lua [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+	append(win, '$', 'range: ' .. (range ~= nil and
+				('['..range.start..', '..range.finish..']')
+				or "invalid range"))
+	append(win, '$', '')
+	return true;
+end)
+
+vis:command('foo')
+vis:command('foo!')
+-- TODO Providing a range does not work
```
Comment (martanne) [inline lua/basic_command_register.lua]: What makes you think it doesn't work? The given range (as all ranges for now) is in byte offsets from the start of the file

File: lua/basic_command_register.lua [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+vis:command('2,4foo!')
+vis:command('%foo')
+vis:command('foo one')
+vis:command('foo one two')
+-- TODO Is there a limit on number of arguments?
+vis:command('foo one two three four five six seven eight nine ten')
```
Comment (martanne) [inline lua/basic_command_register.lua]: Yes, the [maximal number of arguments](https://github.com/martanne/vis/blob/master/sam.c#L24) usable for user code is currently 6, we could probably slightly increase it.

File: lua/test.sh [hunk 3]
```diff
@@ -12,7 +12,7 @@ export VIS_THEME=theme
 if [ $# -gt 0 ]; then
 	test_files=$@
 else
-	test_files=$(find . -type f -name "*.in")
+	test_files="$(find . -type f -name "*.in") basic_empty_file.in"
```
Comment (martanne) [inline lua/test.sh:15]: why?

File: lua/basic_command_register.lua (hunk 1 above)
Comment (joshaw) [inline lua/basic_command_register.lua]: Sorry, some of these fixes are WIP and I was going to come back to them. This one was just a note to say that I wasn't getting what I expected, but in not entirely sure why yet.

File: lua/test.sh (hunk 3 above)
Comment (joshaw) [inline lua/test.sh:15]: I wanted a way to run some tests on a file that doesn't exist. Perhaps there's a better way.

File: lua/basic_command_register.lua (hunk 2 above)
Comment (joshaw) [inline lua/basic_command_register.lua]: Thanks, this was a note to myself to go and check the source, thanks for pointing it out. Again, I know the api is still changing regularly, but as it starts to settle down are you in favor of some more explicit documentation?

File: lua/test.sh (hunk 3 above)
Comment (martanne) [inline lua/test.sh:15]: Just add an empty `basic_empty_file.in` to the repository, i.e. `touch basic_empty_file.in`.

Are you going to update these commits?

File: lua/basic_command_register.lua (hunk 2 above)
Comment (martanne) [inline lua/basic_command_register.lua]: Yes, of course.

File: lua/test.sh (hunk 3 above)
Comment (joshaw) [inline lua/test.sh:15]: I meant the distinction between an empty file and one that doesn't exist, `vis file` when file doesn't exist is different to when file is empty.

I will update, but I won't have opportunity till Monday.

File: lua/basic_command_register.lua (hunk 1 above)
Comment (joshaw) [inline lua/basic_command_register.lua]: Sorry, you are right, the range works (as documented, it returns ranges as bytes), but it looks like the cursor position from `cursor.pos` is not always correct. If I do a `vis.win.cursor:to(1,1)` before one of the tests, the position reported by the test does not change. Am I using it wrong?

**Comment by martanne** [comment]
I still did not have time to properly look at what the tests actually do. However because this is a separate repository and you are most likely the only user of the Lua API on this planet I merged your changes anyway.

A couple of things I noticed:
- having only 1 type of test is nice. However it needs some care to also update the corresponding `*.ref` file when changing a test (`basic_file.ref` did not seem to match the number of tests in `basic_file.lua`).
- test should always terminate! This wasn't the case for the `open` test, I removed it for now.
- in general tests proposed for merging should always pass. Due to the separated repository structure this will require some care in the future to simultaneously update both the C code and the test suite. It was unfortunate that this pull requests contained both worthwhile architectural changes and failing test cases.
- `test.sh` lost the executable rights again, it would be nice if you could make sure it keeps it this time

Thanks for your contributions and feedback about the Lua API!

File: lua/open.lua [hunk 4]
```diff
@@ -0,0 +1,12 @@
+local win = vis.win
+local results = {}
+results[1] = win.file.name == 'open.in'
+vis:open('open.aux')
+-- vis:command('e open.aux')
+results[2] = win.file.name == 'open.aux'
```
Comment (martanne) [inline lua/open.lua:6]: `vis:open` currently creates a new window, as `:open` would do. However `win` will still refer to the old/previous window. If you want to access the new one use `vis.win` which should now point to the newly opened.

**Comment by joshaw** [comment]
Thank you again for your feedback, and for accepting the changes. Sorry for the executable bit changes to test.sh, I'm working in cygwin under windows and thought I'd got it respecting the permissions I set, but obviously not. I'll make sure its consistent next time.

**Pull request closed without merging**